third_party/amd/lib/TritonAMDGPUToLLVM/DotOpToLLVM/MFMA.cpp

```diff
@@ -430,6 +460,284 @@ struct DotOpMFMAConversionHelper {
+            // in this case operands will be duplicated:
+            // A' = [1,2,1,2] B' = [3,4,3,4]
+            // C' = (1*3+2*4) + (1*3+2*4) = 22
+            //
+            // Following code adjusts accumulator values in such cases.
+            // If accumulator is integer, shift accumulator right by
+            // log2(duplicationRate). If accumulator is float, multiply accum
+            // with 1/duplicationRate constant.
+            if (kDimInstrSize > kDimOperandSize) {
+              assert(kDimInstrSize % kDimOperandSize == 0);
+              int duplicationRate = kDimInstrSize / kDimOperandSize;
+              assert(llvm::isPowerOf2_32(duplicationRate));
+              if (dstElemTy.isInteger()) {
+                auto shiftSize = llvm::Log2_32(duplicationRate);
+                assert(!accElem.getType().isUnsignedInteger() &&
+                       "MFMA uses signed accumulator");
+                accElem = tb.ashr(accElem, tb.i32_val(shiftSize));
+              } else {
+                auto multiplierAttr =
+                    rewriter.getFloatAttr(dstElemTy, 1.0 / duplicationRate);
+                auto multiplierVal = rewriter.create<LLVM::ConstantOp>(
+                    loc, dstElemTy, multiplierAttr);
+                accElem = tb.fmul(accElem, multiplierVal);
+              }
+            }
+            auto linearIdx = b * numRepM * numRepN * elemsPerVec +
+                             m * numRepN * elemsPerVec + n * elemsPerVec + v;
+            fc[linearIdx] = accElem;
+          }
+        }
+      }
+    }
+
+    // Originally, setprio (high) is set to the high-level dot op. After dot is
+    // being lowered to the series of mfma operations, it should be moved next
+    // to the first mfma leaving the first mfma staying at the low priority. In
+    // this way, incoming warp can be effectively waiting on the first mfma
+    // instruction (low priority) while the other warp is executing mfma with
+    // high priority. Otherwise, incoming warp can break the cluster.
+    if (setPrioOp && firstMfma)
+      setPrioOp->moveAfter(firstMfma.getDefiningOp());
+
+    // replace with new packed result
+    Type structTy = LLVM::LLVMStructType::getLiteral(
+        ctx, SmallVector<Type>(fc.size(), dstElemTy));
+    Value res = packLLElements(loc, typeConverter, fc, rewriter, structTy);
+
+    Type elemtTy = elemTyA;
+    const size_t mmaCount =
+        numRepB * numRepM * numRepN * numRepK * kWidth / kBase;
+    setNumGeneratedMMAs(op, mmaCount, maybeMfmaInsn->getMDim(),
+                        maybeMfmaInsn->getNDim(), maybeMfmaInsn->getKDim(),
+                        elemtTy);
+
+    rewriter.replaceOp(op, res);
+
+    return success();
+  }
+
+  /// Extract vector from rawElems based on kWidth and kBase
+  /// rawElems is a vector of kWidth elements. We need to prepare vector(s) of
+  /// kBase elements for each mfma instruction
+  SmallVector<Value> extractOperands(Value rawElems, int kWidth, int kBase,
```

Comment:
This is just the same as the normal dot's impl?

Reply:
Discussing in the next comment.